main.go

```diff
@@ -79,52 +77,34 @@ func main() {
 		peer.AddSelfDetails()
 	}
 
-	// Start listening for incoming RPC requests from other peers
-	err = server.StartListener()
-	if err != nil {
-		log.Fatal("Could not register RPC listener. Aborting")
-	}
+	super := initGD2Supervisor()
+	go super.ServeBackground()
+
+	super.Add(servers.New())
 
 	sigCh := make(chan os.Signal)
 	signal.Notify(sigCh)
-	go func() {
-		for s := range sigCh {
-			log.WithField("signal", s).Debug("Signal recieved")
-			switch s {
-			case os.Interrupt:
-				log.WithField("signal", s).Info("Recieved SIGTERM. Stopping GlusterD.")
-				gdctx.Rest.Stop()
-				etcdmgmt.DestroyEmbeddedEtcd()
-				server.StopServer()
-				log.Info("Termintaing GlusterD.")
-				os.Exit(0)
-
-			default:
-				continue
-			}
+	for s := range sigCh {
+		log.WithField("signal", s).Debug("Signal recieved")
+		switch s {
+		case os.Interrupt:
+			log.WithField("signal", s).Info("Recieved SIGTERM. Stopping GlusterD.")
```

Comment:
I had the opposite experience. When the signal handler was in a goroutine it didn't catch signals. Having the main thread do it worked for me. I'll need to check this further.